### DISCUSSION TITLE
关于MySQL的多CPU访问

### DISCUSSION THREAD

**Discussion by ericzhanchina** [cloudwu/skynet#1763] [General]
目前是使用了一个单独的Service来访问MySQL的，MySQL由于是多核的，我理解建立多个链接的话效率肯定会更高，连接数与核心数保持一致。然后这个service维持着多个mysql的链接，保证每个服务的访问都提交到同一个链接上，比如使用一种简单的hash算法。

是不是这样的方式就可以充分利用多核的优势了，请教。

**Comment by terry8210** [comment]
应该多开几个service，然后外层使用proxy的方式，通过pid进行hash找到对应的mysql service进行数据库操作。单个服务包含多个mysql链接，发挥不了。

**Reply by ericzhanchina** [reply]
我之前是你这样做的，但是云风说这样做不对的，应该单个服务开多个链接

**Reply by terry8210** [reply]
云风怎么说的?如果一个服务发起多个链接，意义不大哦。

**Reply by ericzhanchina** [reply]
刚才找了一下之前的帖子没找到，今天再翻翻，还得云风老大给解释一下才好心里踏实

**Reply by cloudwu** [reply]
拆分 service 发多个连接没什么意义，但是分连接本身应该是有意义的，因为：

1. mysql 那边可以并行处理。
2. 如果单个请求的体积很大的话，多个连接不会阻塞住其它请求。

如果在单个服务做，注意也要用 coroutine 分开，而不要串行处理。这样，在一个请求太大被阻塞时，不至于阻塞其它请求。如果回应数据需要二次加工，可以考虑分 service 。

但我觉得整个系统不应该把 cpu 负担加在 mysql 上，而是尽量把业务逻辑放在 skynet 里面。mysql 只是做简单查询的话，无论怎样做都不会成为 cpu 瓶颈。至于 skynet 到 mysql 的 IO 吞吐量，不会因为增加连接而改善。

**Reply by ericzhanchina** [reply]
后来我想了一下，因为Game服已经设计成分布式的了，横向扩展，那么每个Game服跟mysql建立一个连接，4个Game服也建立了四个连接，我想也是充分利用了mysql的多连接处理的优势。

**Comment by CoolDesert** [comment]
skynet 只有一个 io 线程，即使是多条连接，最终还是汇总到一起。所以拆分开意义不大。
